File: desktop/src-tauri/src/auto_install/schema.rs
use tauri::AppHandle;

use super::errors::format_worker_http_error;
use super::log::emit_log;
use super::types::InitDbResult;

fn is_transient_schema_error(err: &str) -> bool {
    let lower = err.to_lowercase();
    lower.contains("1042")
        || lower.contains("404")
        || lower.contains("1101")
        || lower.contains("1104")
        || lower.contains("not found")
        || lower.contains("401")
        || lower.contains("unauthorized")
}

pub(crate) async fn init_worker_db_with_retry(
    app: &AppHandle,
    worker_url: &str,
    pepper: Option<&str>,
    access_token: Option<&str>,
) -> Result<InitDbResult, String> {
    const ATTEMPTS: u32 = 4;
    let mut last = String::new();
    for attempt in 1..=ATTEMPTS {
        match init_worker_db(worker_url, pepper, access_token).await {
            Ok(r) => return Ok(r),
            Err(e) => {
                last = e;
                if attempt == ATTEMPTS || !is_transient_schema_error(&last) {
                    break;
                }
                emit_log(
                    app,
                    "init-db",
                    "info",
                    format!("init-db not ready yet (attempt {attempt}/{ATTEMPTS}) — retrying…"),
                );
                tokio::time::sleep(std::time::Duration::from_secs(2)).await;
            }
        }
    }
    Err(last)
}

pub(crate) async fn migrate_worker_db_with_retry(
    app: &AppHandle,
    worker_url: &str,
    pepper: Option<&str>,
    access_token: Option<&str>,
) -> Result<InitDbResult, String> {
    const ATTEMPTS: u32 = 4;
    let mut last = String::new();
    for attempt in 1..=ATTEMPTS {
        match migrate_worker_db(worker_url, pepper, access_token).await {
            Ok(r) => return Ok(r),
            Err(e) => {
                last = e;
                if attempt == ATTEMPTS || !is_transient_schema_error(&last) {
                    break;
                }
                emit_log(
                    app,
                    "migrate-db",
                    "info",
                    format!("migrate-db not ready yet (attempt {attempt}/{ATTEMPTS}) — retrying…"),
                );
                tokio::time::sleep(std::time::Duration::from_secs(2)).await;
            }
        }
    }
    Err(last)
}

async fn post_schema_endpoint(
    worker_url: &str,
    pepper: Option<&str>,
    access_token: Option<&str>,
    path: &str,
    step: &str,
) -> Result<InitDbResult, String> {
    let base = worker_url.trim().trim_end_matches('/');
    if base.is_empty() {
        return Err("Worker URL is empty".into());
    }
    let url = format!("{base}{path}");
    let client = reqwest::Client::new();
    let mut req = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&serde_json::json!({}));
    if let Some(token) = access_token.filter(|s| !s.trim().is_empty()) {
        req = req.header("Authorization", format!("Bearer {}", token.trim()));
    } else if let Some(p) = pepper.filter(|s| !s.trim().is_empty()) {
        req = req.header("X-Auth-Pepper", p.trim());
    } else {
        return Err(format!("{step} requires an owner session or AUTH_PEPPER"));
    }
    let res = req
        .send()
        .await
        .map_err(|e| format!("{step} request failed: {e}"))?;
    if !res.status().is_success() {
        let status = res.status();
        let text = res.text().await.unwrap_or_default();
        return Err(format_worker_http_error(step, status, &text));
    }
    res.json::<InitDbResult>()
        .await
        .map_err(|e| format!("{step} response parse failed: {e}"))
}

/// Empty D1 only. Fails with 409 if product tables already exist.
pub async fn init_worker_db(
    worker_url: &str,
    pepper: Option<&str>,
    access_token: Option<&str>,
) -> Result<InitDbResult, String> {
    post_schema_endpoint(worker_url, pepper, access_token, "/console/init-db", "init-db").await
}

/// Pending migrations only. Never drops tables.
pub async fn migrate_worker_db(
    worker_url: &str,
    pepper: Option<&str>,
    access_token: Option<&str>,
) -> Result<InitDbResult, String> {
    post_schema_endpoint(worker_url, pepper, access_token, "/console/migrate-db", "migrate-db").await
}

File: app/src/server/cloudflare/schema.ts
// Calls the deployed Worker's schema + owner-setup endpoints — ported from
// desktop/src-tauri/src/auto_install/{schema,errors}.rs and
// desktop/src-tauri/src/auth/owner_session.rs (owner_setup_admin).

export type InitDbResult = {
  ok: boolean;
  alreadyInitialized: boolean;
  applied: string[];
  skipped: string[];
  cleared: boolean;
};

function isTransientSchemaError(err: string): boolean {
  const lower = err.toLowerCase();
  return (
    lower.includes("1042") ||
    lower.includes("404") ||
    lower.includes("1101") ||
    lower.includes("1104") ||
    lower.includes("not found") ||
    lower.includes("401") ||
    lower.includes("unauthorized")
  );
}

async function postSchemaEndpoint(
  workerUrl: string,
  pepper: string | undefined,
  path: string,
  step: string,
  cfAccessToken?: string,
): Promise<InitDbResult> {
  const base = workerUrl.trim().replace(/\/$/, "");
  if (!base) throw new Error("Worker URL is empty");
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (pepper?.trim()) headers["X-Auth-Pepper"] = pepper.trim();
  else if (cfAccessToken?.trim()) headers["X-Cf-Access-Token"] = cfAccessToken.trim();
  else throw new Error(`${step} requires AUTH_PEPPER or Cloudflare OAuth`);
  const res = await fetch(`${base}${path}`, { method: "POST", headers, body: "{}" });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`${step} returned ${res.status}: ${text.slice(0, 280)}`);
  }
  const value = await res.json();
  return {
    ok: Boolean(value.ok),
    alreadyInitialized: Boolean(value.alreadyInitialized ?? value.already_initialized),
    applied: value.applied ?? [],
    skipped: value.skipped ?? [],
    cleared: Boolean(value.cleared),
  };
}

async function withRetry(
  fn: () => Promise<InitDbResult>,
  step: string,
  onLog?: (line: string) => void,
): Promise<InitDbResult> {
  const attempts = 4;
  let lastErr: unknown;
  for (let attempt = 1; attempt <= attempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (attempt === attempts || !isTransientSchemaError(String(err))) break;
      onLog?.(`${step} not ready yet (attempt ${attempt}/${attempts}) — retrying…`);
      await new Promise((resolve) => setTimeout(resolve, 2000));
    }
  }
  throw lastErr;
}

/** Empty D1 only. Fails with 409 if product tables already exist. */
export async function initWorkerDb(
  workerUrl: string,
  pepper: string | undefined,
  onLog?: (line: string) => void,
): Promise<InitDbResult> {
  return withRetry(
    () => postSchemaEndpoint(workerUrl, pepper, "/console/init-db", "init-db"),
    "init-db",
    onLog,
  );
}

/** Pending migrations only. Never drops tables. */
export async function migrateWorkerDb(
  workerUrl: string,
  pepper: string | undefined,
  onLog?: (line: string) => void,
  cfAccessToken?: string,
): Promise<InitDbResult> {
  return withRetry(
    () =>
      postSchemaEndpoint(
        workerUrl,
        pepper,
        "/console/migrate-db",
        "migrate-db",
        cfAccessToken,
      ),
    "migrate-db",
    onLog,
  );
}

export type OwnerSetupResult = { passtoken: string };

/** POST /console/setup-admin with X-Auth-Pepper — issues the one-time owner passtoken. */
export async function ownerSetupAdmin(
  workerUrl: string,
  pepper: string,
): Promise<OwnerSetupResult> {
  const base = workerUrl.trim().replace(/\/$/, "");
  if (!base) throw new Error("Worker URL is required");
  if (!pepper.trim()) throw new Error("Install pepper is missing. Re-run install.");
  const res = await fetch(`${base}/console/setup-admin`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "X-Auth-Pepper": pepper.trim() },
    body: "{}",
  });
  const value = await res.json().catch(() => ({}));
  if (res.status !== 200 && res.status !== 201) {
    throw new Error(value?.error ?? "Could not set up owner");
  }
  const passtoken = value?.passtoken;
  if (!passtoken) throw new Error("Worker did not return a passtoken");
  return { passtoken };
}

/** POST /console/reset-admin — re-issue passtoken when setup-admin cannot run. */
export async function resetOwnerAdmin(
  workerUrl: string,
  cfAccessToken: string,
  cfAccountId?: string,
): Promise<OwnerSetupResult> {
  const base = workerUrl.trim().replace(/\/$/, "");
  if (!base) throw new Error("Worker URL is required");
  const token = cfAccessToken.trim();
  if (!token) throw new Error("Cloudflare access token is required");
  const res = await fetch(`${base}/console/reset-admin`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      cfAccessToken: token,
      ...(cfAccountId?.trim() ? { cfAccountId: cfAccountId.trim() } : {}),
    }),
  });
  const value = (await res.json().catch(() => ({}))) as {
    passtoken?: string;
    error?: string;
  };
  if (!res.ok || !value.passtoken) {
    throw new Error(value.error ?? "Could not re-issue owner passtoken");
  }
  return { passtoken: value.passtoken };
}

export async function waitForWorkerReady(
  workerUrl: string,
  onLog?: (line: string) => void,
): Promise<void> {
  const backoffs = [2, 4, 8, 16];
  onLog?.(`Waiting for ${workerUrl} to become reachable…`);
  for (let attempt = 0; attempt <= backoffs.length; attempt++) {
    if (attempt > 0) {
      const delay = backoffs[attempt - 1];
      onLog?.(`Worker not ready yet — retrying in ${delay}s (attempt ${attempt + 1}/${backoffs.length + 1})…`);
      await new Promise((resolve) => setTimeout(resolve, delay * 1000));
    }
    try {
      const res = await fetch(`${workerUrl.replace(/\/$/, "")}/health`);
      if (res.ok) {
        const json = await res.json().catch(() => ({}));
        if (json?.ok === true) {
          onLog?.(`Worker is reachable (attempt ${attempt + 1})`);
          return;
        }
      }
    } catch {
      /* retry */
    }
  }
  onLog?.("Worker did not respond to /health within ~30s — continuing anyway.");
}

/** Public probe: is an owner passtoken already configured on this Worker? */
export async function fetchOwnerConfigured(workerUrl: string): Promise<boolean> {
  try {
    const res = await fetch(`${workerUrl.replace(/\/$/, "")}/console/auth-status`);
    if (!res.ok) return false;
    const json = await res.json();
    return Boolean(json?.ownerConfigured);
  } catch {
    return false;
  }
}

export async function fetchWorkerVersion(workerUrl: string): Promise<string | null> {
  try {
    const res = await fetch(`${workerUrl.replace(/\/$/, "")}/health`);
    if (!res.ok) return null;
    const json = await res.json();
    const version = String(json?.version ?? "").trim();
    return version && version !== "unknown" ? version : null;
  } catch {
    return null;
  }
}
